File: sdks/rust/tests/bitmap_s2s_test.rs
//! Bitmap S2S Integration Tests
//!
//! These tests require a running Synap server.
//! Run with: SYNAP_URL=http://localhost:15500 cargo test --test bitmap_s2s_test

mod common;

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_setbit_getbit() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Set bit 5 to 1
    let old_value = client.bitmap().setbit(&key, 5, 1).await.unwrap();
    assert_eq!(old_value, 0); // Was unset before

    // Get bit 5
    let value = client.bitmap().getbit(&key, 5).await.unwrap();
    assert_eq!(value, 1);

    // Set bit 5 back to 0
    let old_value2 = client.bitmap().setbit(&key, 5, 0).await.unwrap();
    assert_eq!(old_value2, 1); // Was set before

    // Get bit 5 again
    let value2 = client.bitmap().getbit(&key, 5).await.unwrap();
    assert_eq!(value2, 0);
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitcount() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:count:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Set multiple bits
    client.bitmap().setbit(&key, 0, 1).await.unwrap();
    client.bitmap().setbit(&key, 2, 1).await.unwrap();
    client.bitmap().setbit(&key, 4, 1).await.unwrap();
    client.bitmap().setbit(&key, 6, 1).await.unwrap();

    // Count all bits
    let count = client.bitmap().bitcount(&key, None, None).await.unwrap();
    assert_eq!(count, 4);
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitpos() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:pos:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Set bit at position 7
    client.bitmap().setbit(&key, 7, 1).await.unwrap();

    // Find first set bit
    let pos = client.bitmap().bitpos(&key, 1, None, None).await.unwrap();
    assert_eq!(pos, Some(7));
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitop_and() {
    let client = common::setup_s2s_client();
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let key1 = format!("test:bitmap:and1:{}", timestamp);
    let key2 = format!("test:bitmap:and2:{}", timestamp);
    let dest = format!("test:bitmap:and_result:{}", timestamp);

    // Set bits in bitmap1 (bits 0, 1, 2)
    client.bitmap().setbit(&key1, 0, 1).await.unwrap();
    client.bitmap().setbit(&key1, 1, 1).await.unwrap();
    client.bitmap().setbit(&key1, 2, 1).await.unwrap();

    // Set bits in bitmap2 (bits 1, 2, 3)
    client.bitmap().setbit(&key2, 1, 1).await.unwrap();
    client.bitmap().setbit(&key2, 2, 1).await.unwrap();
    client.bitmap().setbit(&key2, 3, 1).await.unwrap();

    // AND operation
    let length = client
        .bitmap()
        .bitop(
            synap_sdk::BitmapOperation::And,
            &dest,
            &[key1.clone(), key2.clone()],
        )
        .await
        .unwrap();
    assert!(length > 0);

    // Check result: should have bits 1 and 2 set
    assert_eq!(client.bitmap().getbit(&dest, 0).await.unwrap(), 0);
    assert_eq!(client.bitmap().getbit(&dest, 1).await.unwrap(), 1);
    assert_eq!(client.bitmap().getbit(&dest, 2).await.unwrap(), 1);
    assert_eq!(client.bitmap().getbit(&dest, 3).await.unwrap(), 0);
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitfield_get_set() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:bitfield:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // SET operation: Set 8-bit unsigned value 42 at offset 0
    let operations = vec![serde_json::json!({
        "operation": "SET",
        "offset": 0,
        "width": 8,
        "signed": false,
        "value": 42
    })];

    let results = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results.len(), 1);
    assert_eq!(results[0], 0); // Old value was 0

    // GET operation: Read back the value
    let operations = vec![serde_json::json!({
        "operation": "GET",
        "offset": 0,
        "width": 8,
        "signed": false
    })];

    let results = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results.len(), 1);
    assert_eq!(results[0], 42);
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitfield_incrby_wrap() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:bitfield_wrap:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Set initial value
    let operations = vec![serde_json::json!({
        "operation": "SET",
        "offset": 0,
        "width": 8,
        "signed": false,
        "value": 250
    })];
    client.bitmap().bitfield(&key, &operations).await.unwrap();

    // INCRBY with wrap: 250 + 10 = 260 wraps to 4
    let operations = vec![serde_json::json!({
        "operation": "INCRBY",
        "offset": 0,
        "width": 8,
        "signed": false,
        "increment": 10,
        "overflow": "WRAP"
    })];

    let results = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results.len(), 1);
    assert_eq!(results[0], 4); // 250 + 10 = 260 wraps to 4 (260 - 256)
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitfield_incrby_sat() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:bitfield_sat:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Set 4-bit unsigned value to 14
    let operations = vec![serde_json::json!({
        "operation": "SET",
        "offset": 0,
        "width": 4,
        "signed": false,
        "value": 14
    })];
    client.bitmap().bitfield(&key, &operations).await.unwrap();

    // INCRBY with saturate: 14 + 1 = 15 (max), then stays at 15
    let operations = vec![serde_json::json!({
        "operation": "INCRBY",
        "offset": 0,
        "width": 4,
        "signed": false,
        "increment": 1,
        "overflow": "SAT"
    })];

    let results1 = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results1[0], 15);

    // Try to increment again (should saturate at 15)
    let results2 = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results2[0], 15);
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitfield_multiple_operations() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:bitfield_multi:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Execute multiple operations in sequence
    let operations = vec![
        serde_json::json!({
            "operation": "SET",
            "offset": 0,
            "width": 8,
            "signed": false,
            "value": 100
        }),
        serde_json::json!({
            "operation": "SET",
            "offset": 8,
            "width": 8,
            "signed": false,
            "value": 200
        }),
        serde_json::json!({
            "operation": "GET",
            "offset": 0,
            "width": 8,
            "signed": false
        }),
        serde_json::json!({
            "operation": "GET",
            "offset": 8,
            "width": 8,
            "signed": false
        }),
        serde_json::json!({
            "operation": "INCRBY",
            "offset": 0,
            "width": 8,
            "signed": false,
            "increment": 50,
            "overflow": "WRAP"
        }),
    ];

    let results = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results.len(), 5);
    assert_eq!(results[0], 0); // Old value at offset 0
    assert_eq!(results[1], 0); // Old value at offset 8
    assert_eq!(results[2], 100); // Read back offset 0
    assert_eq!(results[3], 200); // Read back offset 8
    assert_eq!(results[4], 150); // Incremented offset 0
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_bitfield_signed_values() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:bitfield_signed:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Set signed 8-bit negative value
    let operations = vec![serde_json::json!({
        "operation": "SET",
        "offset": 0,
        "width": 8,
        "signed": true,
        "value": -10
    })];
    client.bitmap().bitfield(&key, &operations).await.unwrap();

    // Read back as signed
    let operations = vec![serde_json::json!({
        "operation": "GET",
        "offset": 0,
        "width": 8,
        "signed": true
    })];

    let results = client.bitmap().bitfield(&key, &operations).await.unwrap();
    assert_eq!(results.len(), 1);
    assert_eq!(results[0], -10);
}

#[tokio::test]
#[ignore = "requires running Synap server"]
async fn test_bitmap_stats() {
    let client = common::setup_s2s_client();
    let key = format!(
        "test:bitmap:stats:{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    );

    // Perform some operations
    client.bitmap().setbit(&key, 0, 1).await.unwrap();
    client.bitmap().getbit(&key, 0).await.unwrap();
    client.bitmap().bitcount(&key, None, None).await.unwrap();

    let stats = client.bitmap().stats().await.unwrap();
    assert!(stats.setbit_count >= 1);
    assert!(stats.getbit_count >= 1);
    assert!(stats.bitcount_count >= 1);
}

File: sdks/typescript/src/__tests__/bitmap.s2s.test.ts
/**
 * Bitmap Manager S2S (Server-to-Server) Tests
 * 
 * Integration tests that require a running Synap server.
 * These tests are skipped if SYNAP_URL is not set.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import { Synap } from '../index';

const SERVER_URL = process.env.SYNAP_URL || 'http://localhost:15500';
const SKIP_S2S = !process.env.SYNAP_URL && !process.env.CI;

describe.skipIf(SKIP_S2S)('BitmapManager S2S', () => {
  let synap: Synap;

  beforeAll(() => {
    synap = new Synap({ url: SERVER_URL });
  });

  describe('SETBIT/GETBIT', () => {
    it('should set and get bits', async () => {
      const key = `test:bitmap:${Date.now()}`;

      // Set bit 5 to 1
      const oldValue1 = await synap.bitmap.setbit(key, 5, 1);
      expect(oldValue1).toBe(0); // Was unset before

      // Get bit 5
      const value1 = await synap.bitmap.getbit(key, 5);
      expect(value1).toBe(1);

      // Set bit 5 back to 0
      const oldValue2 = await synap.bitmap.setbit(key, 5, 0);
      expect(oldValue2).toBe(1); // Was set before

      // Get bit 5 again
      const value2 = await synap.bitmap.getbit(key, 5);
      expect(value2).toBe(0);
    });
  });

  describe('BITCOUNT', () => {
    it('should count set bits', async () => {
      const key = `test:bitmap:count:${Date.now()}`;

      // Set multiple bits
      await synap.bitmap.setbit(key, 0, 1);
      await synap.bitmap.setbit(key, 2, 1);
      await synap.bitmap.setbit(key, 4, 1);
      await synap.bitmap.setbit(key, 6, 1);

      // Count all bits
      const count = await synap.bitmap.bitcount(key);
      expect(count).toBe(4);
    });

    it('should count bits in range', async () => {
      const key = `test:bitmap:range:${Date.now()}`;

      // Set bits at positions 0, 1, 2, 10, 11, 12
      await synap.bitmap.setbit(key, 0, 1);
      await synap.bitmap.setbit(key, 1, 1);
      await synap.bitmap.setbit(key, 2, 1);
      await synap.bitmap.setbit(key, 10, 1);
      await synap.bitmap.setbit(key, 11, 1);
      await synap.bitmap.setbit(key, 12, 1);

      // Count in range [0, 7] (first byte)
      const count1 = await synap.bitmap.bitcount(key, 0, 7);
      expect(count1).toBe(3);

      // Count in range [8, 15] (second byte)
      const count2 = await synap.bitmap.bitcount(key, 8, 15);
      expect(count2).toBe(3);
    });
  });

  describe('BITPOS', () => {
    it('should find position of set bit', async () => {
      const key = `test:bitmap:pos:${Date.now()}`;

      // Set bit at position 7
      await synap.bitmap.setbit(key, 7, 1);

      // Find first set bit
      const pos = await synap.bitmap.bitpos(key, 1);
      expect(pos).toBe(7);
    });

    it('should find position with start offset', async () => {
      const key = `test:bitmap:pos2:${Date.now()}`;

      // Set bits at 3 and 10
      await synap.bitmap.setbit(key, 3, 1);
      await synap.bitmap.setbit(key, 10, 1);

      // Find from start
      const pos1 = await synap.bitmap.bitpos(key, 1);
      expect(pos1).toBe(3);

      // Find from offset 5
      const pos2 = await synap.bitmap.bitpos(key, 1, 5);
      expect(pos2).toBe(10);
    });

    it('should return null when bit not found', async () => {
      const key = `test:bitmap:pos3:${Date.now()}`;

      // Empty bitmap - server may return error or null
      try {
        const pos = await synap.bitmap.bitpos(key, 1);
        // If it succeeds, should be null for empty bitmap
        expect(pos).toBeNull();
      } catch (error: any) {
        // Server may return error for non-existent key, which is also acceptable
        expect(error).toBeDefined();
      }
    });
  });

  describe('BITOP', () => {
    it('should perform AND operation', async () => {
      const key1 = `test:bitmap:and1:${Date.now()}`;
      const key2 = `test:bitmap:and2:${Date.now()}`;
      const dest = `test:bitmap:and_result:${Date.now()}`;

      // Set bits in bitmap1 (bits 0, 1, 2)
      await synap.bitmap.setbit(key1, 0, 1);
      await synap.bitmap.setbit(key1, 1, 1);
      await synap.bitmap.setbit(key1, 2, 1);

      // Set bits in bitmap2 (bits 1, 2, 3)
      await synap.bitmap.setbit(key2, 1, 1);
      await synap.bitmap.setbit(key2, 2, 1);
      await synap.bitmap.setbit(key2, 3, 1);

      // AND operation
      const length = await synap.bitmap.bitop('AND', dest, [key1, key2]);
      expect(length).toBeGreaterThan(0);

      // Check result: should have bits 1 and 2 set
      expect(await synap.bitmap.getbit(dest, 0)).toBe(0);
      expect(await synap.bitmap.getbit(dest, 1)).toBe(1);
      expect(await synap.bitmap.getbit(dest, 2)).toBe(1);
      expect(await synap.bitmap.getbit(dest, 3)).toBe(0);
    });

    it('should perform OR operation', async () => {
      const key1 = `test:bitmap:or1:${Date.now()}`;
      const key2 = `test:bitmap:or2:${Date.now()}`;
      const dest = `test:bitmap:or_result:${Date.now()}`;

      await synap.bitmap.setbit(key1, 0, 1);
      await synap.bitmap.setbit(key2, 1, 1);

      const length = await synap.bitmap.bitop('OR', dest, [key1, key2]);
      expect(length).toBeGreaterThan(0);

      expect(await synap.bitmap.getbit(dest, 0)).toBe(1);
      expect(await synap.bitmap.getbit(dest, 1)).toBe(1);
    });

    it('should perform XOR operation', async () => {
      const key1 = `test:bitmap:xor1:${Date.now()}`;
      const key2 = `test:bitmap:xor2:${Date.now()}`;
      const dest = `test:bitmap:xor_result:${Date.now()}`;

      await synap.bitmap.setbit(key1, 0, 1);
      await synap.bitmap.setbit(key1, 1, 1);
      await synap.bitmap.setbit(key2, 1, 1);

      const length = await synap.bitmap.bitop('XOR', dest, [key1, key2]);
      expect(length).toBeGreaterThan(0);

      // XOR: both set = 0, one set = 1
      expect(await synap.bitmap.getbit(dest, 0)).toBe(1);
      expect(await synap.bitmap.getbit(dest, 1)).toBe(0);
    });

    it('should perform NOT operation', async () => {
      const key = `test:bitmap:not1:${Date.now()}`;
      const dest = `test:bitmap:not_result:${Date.now()}`;

      await synap.bitmap.setbit(key, 0, 1);
      await synap.bitmap.setbit(key, 2, 1);

      const length = await synap.bitmap.bitop('NOT', dest, [key]);
      expect(length).toBeGreaterThan(0);

      // NOT: 1 becomes 0, 0 becomes 1
      expect(await synap.bitmap.getbit(dest, 0)).toBe(0);
      expect(await synap.bitmap.getbit(dest, 1)).toBe(1);
      expect(await synap.bitmap.getbit(dest, 2)).toBe(0);
    });
  });

  describe('BITFIELD', () => {
    it('should execute GET and SET operations', async () => {
      const key = `test:bitmap:bitfield:${Date.now()}`;

      // SET operation: Set 8-bit unsigned value 42 at offset 0
      const setResults = await synap.bitmap.bitfield(key, [
        {
          operation: 'SET',
          offset: 0,
          width: 8,
          signed: false,
          value: 42,
        },
      ]);

      expect(setResults).toHaveLength(1);
      expect(setResults[0]).toBe(0); // Old value was 0

      // GET operation: Read back the value
      const getResults = await synap.bitmap.bitfield(key, [
        {
          operation: 'GET',
          offset: 0,
          width: 8,
          signed: false,
        },
      ]);

      expect(getResults).toHaveLength(1);
      expect(getResults[0]).toBe(42);
    });

    it('should execute INCRBY with WRAP overflow', async () => {
      const key = `test:bitmap:bitfield_wrap:${Date.now()}`;

      // Set initial value
      await synap.bitmap.bitfield(key, [
        {
          operation: 'SET',
          offset: 0,
          width: 8,
          signed: false,
          value: 250,
        },
      ]);

      // INCRBY with wrap: 250 + 10 = 260 wraps to 4
      const results = await synap.bitmap.bitfield(key, [
        {
          operation: 'INCRBY',
          offset: 0,
          width: 8,
          signed: false,
          increment: 10,
          overflow: 'WRAP',
        },
      ]);

      expect(results).toHaveLength(1);
      expect(results[0]).toBe(4); // 250 + 10 = 260 wraps to 4 (260 - 256)
    });

    it('should execute INCRBY with SAT overflow', async () => {
      const key = `test:bitmap:bitfield_sat:${Date.now()}`;

      // Set 4-bit unsigned value to 14
      await synap.bitmap.bitfield(key, [
        {
          operation: 'SET',
          offset: 0,
          width: 4,
          signed: false,
          value: 14,
        },
      ]);

      // INCRBY with saturate: 14 + 1 = 15 (max), then stays at 15
      const results1 = await synap.bitmap.bitfield(key, [
        {
          operation: 'INCRBY',
          offset: 0,
          width: 4,
          signed: false,
          increment: 1,
          overflow: 'SAT',
        },
      ]);

      expect(results1[0]).toBe(15);

      // Try to increment again (should saturate at 15)
      const results2 = await synap.bitmap.bitfield(key, [
        {
          operation: 'INCRBY',
          offset: 0,
          width: 4,
          signed: false,
          increment: 1,
          overflow: 'SAT',
        },
      ]);

      expect(results2[0]).toBe(15);
    });

    it('should execute multiple operations', async () => {
      const key = `test:bitmap:bitfield_multi:${Date.now()}`;

      // Execute multiple operations in sequence
      const results = await synap.bitmap.bitfield(key, [
        {
          operation: 'SET',
          offset: 0,
          width: 8,
          signed: false,
          value: 100,
        },
        {
          operation: 'SET',
          offset: 8,
          width: 8,
          signed: false,
          value: 200,
        },
        {
          operation: 'GET',
          offset: 0,
          width: 8,
          signed: false,
        },
        {
          operation: 'GET',
          offset: 8,
          width: 8,
          signed: false,
        },
        {
          operation: 'INCRBY',
          offset: 0,
          width: 8,
          signed: false,
          increment: 50,
          overflow: 'WRAP',
        },
      ]);

      expect(results).toHaveLength(5);
      expect(results[0]).toBe(0); // Old value at offset 0
      expect(results[1]).toBe(0); // Old value at offset 8
      expect(results[2]).toBe(100); // Read back offset 0
      expect(results[3]).toBe(200); // Read back offset 8
      expect(results[4]).toBe(150); // Incremented offset 0
    });

    it('should handle signed values', async () => {
      const key = `test:bitmap:bitfield_signed:${Date.now()}`;

      // Set signed 8-bit negative value
      await synap.bitmap.bitfield(key, [
        {
          operation: 'SET',
          offset: 0,
          width: 8,
          signed: true,
          value: -10,
        },
      ]);

      // Read back as signed
      const results = await synap.bitmap.bitfield(key, [
        {
          operation: 'GET',
          offset: 0,
          width: 8,
          signed: true,
        },
      ]);

      expect(results[0]).toBe(-10);
    });
  });

  describe('STATS', () => {
    it('should retrieve bitmap statistics', async () => {
      const key = `test:bitmap:stats:${Date.now()}`;

      // Perform some operations
      await synap.bitmap.setbit(key, 0, 1);
      await synap.bitmap.getbit(key, 0);
      await synap.bitmap.bitcount(key);

      const stats = await synap.bitmap.stats();

      expect(stats).toHaveProperty('total_bitmaps');
      expect(stats).toHaveProperty('total_bits');
      expect(stats).toHaveProperty('setbit_count');
      expect(stats).toHaveProperty('getbit_count');
      expect(stats).toHaveProperty('bitcount_count');
      expect(stats).toHaveProperty('bitop_count');
      expect(stats).toHaveProperty('bitpos_count');
      expect(stats).toHaveProperty('bitfield_count');

      expect(stats.setbit_count).toBeGreaterThanOrEqual(1);
      expect(stats.getbit_count).toBeGreaterThanOrEqual(1);
      expect(stats.bitcount_count).toBeGreaterThanOrEqual(1);
    });
  });
});

